### PR TITLE
 Strongly typed form values for Flow and Typescript

### PR DESCRIPTION
**Pull request by erikras**
People have been attempting to get close to this for a while now.

It's my understanding that there's no good way of typing the actual fields. What this provides is strong typing for the `values` and `initialValues` when using Final Form.

### CODE REVIEW

**Comment by erikras** [comment]
This pull request **introduces 3 alerts** when merging 7781d0b8b6ddf2b24304d301d3194e059f646a03 into 0d3a88600d723976644e3df9073f329b8c513b3d - [view on LGTM.com](https://lgtm.com/projects/g/final-form/final-form/rev/pr-549975e982c0e85d2c80467ab3b68dda0eefc4fe)

**new alerts:**

* 3 for Useless assignment to local variable

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by erikras** [comment]
This pull request **introduces 2 alerts** when merging a3b1370218368911f4493ee9e63e20c399608eac into 0d3a88600d723976644e3df9073f329b8c513b3d - [view on LGTM.com](https://lgtm.com/projects/g/final-form/final-form/rev/pr-79cb48526aaf8f998bbf4ec2c6017acebf965082)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Unused variable, import, function or class

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/index.d.ts [hunk 1]
```diff
@@ -180,73 +180,77 @@ export interface InternalFormState {
...
+  debug?: DebugFunction<FormValues>
   destroyOnUnregister?: boolean
-  initialValues?: FormData
+  initialValues?: FormValues
   keepDirtyOnReinitialize?: boolean
   mutators?: { [key: string]: Mutator }
   onSubmit: (
-    values: FormData,
+    values: FormValues,
     form: FormApi,
```
Comment (ChadLefort) [inline src/index.d.ts]: Can `FormApi` get `FormValues` passed to it also?

Comment (erikras) [reply src/index.d.ts]: Wow. Good catch.

**Comment by erikras** [comment]
This pull request **introduces 2 alerts** when merging cc4a0079449eab1e26d47f2ae1c5fe15360a5666 into 0d3a88600d723976644e3df9073f329b8c513b3d - [view on LGTM.com](https://lgtm.com/projects/g/final-form/final-form/rev/pr-cd83cf7f283efb890815e620c2e76c5553ae4289)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Unused variable, import, function or class

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by erikras** [comment]
Published in [`v4.14.0`](https://github.com/final-form/final-form/releases/tag/v4.14.0).

File: src/index.d.ts [hunk 2]
```diff
@@ -180,91 +180,97 @@ export interface InternalFormState {
...
   destroyOnUnregister?: boolean
-  initialValues?: FormData
+  initialValues?: FormValues
   keepDirtyOnReinitialize?: boolean
   mutators?: { [key: string]: Mutator }
   onSubmit: (
-    values: FormData,
-    form: FormApi,
+    values: FormValues,
+    form?: FormApi<FormValues>,
```
Comment (Jephuff) [inline src/index.d.ts]: @erikras should this be optional? It looks like it's not a maybe type in the flow types. 
https://github.com/final-form/final-form/pull/223/files#diff-63dfa14a27e7a50ad2a9186e1bcc0e48R271

(p.s. thanks for the typescript updates :D )

Comment (erikras) [reply src/index.d.ts]: Good catch. Will fix.

Comment (erikras) [reply src/index.d.ts]: Fixed in [`v4.14.1`](https://github.com/final-form/final-form/releases/tag/v4.14.1).

Comment (Jephuff) [reply src/index.d.ts]: Amazing! Thank you :)

**Pull request merged**